src/node_modules/app/elasticsearch/index.js

```diff
@@ -25,6 +29,72 @@ export function getESType (field) {
 	return 'object';
 }
 
+export function simplifyType (typeObject) {
+	let type = typeObject;
+	if (Array.isArray(type)) {
+		type = type.filter(t => t !== 'undefined')
+		if (type.length === 1) {
+			[type] = type;
+		}
+	}
+	return type;
+}
+
+export function computePresentationType (fullType, indentation = 0) {
+	if (typeof fullType === 'string') {
+		return fullType;
+	}
+	if (Array.isArray(fullType)) {
+		return fullType
+		.filter(type => type !== 'undefined')
+		.map(type => computePresentationType(type, indentation + 1))
+		.join(' | ');
+	}
+	if (typeof fullType === 'object') {
+		if ('__array' in fullType) {
+			return `${computePresentationType(fullType.__array, indentation + 1)}[]`;
+		}
+		if ('__legacyName' in fullType) {
+			return fullType.__legacyName;
+		}
+		const paramNames = Object.keys(fullType);
+		const result = paramNames
+		.map(name => {
+			return `${'\t'.repeat(indentation)}${name}: ${computePresentationType(fullType[name], indentation + 1)};`;
+		})
+		.join('\n');
+		return ['{\n', result, '\n', '\t'.repeat(indentation >= 1 ? indentation - 1 : 0), '}'].join('');
+	}
+	throw new Error('Failed to parse type data, bad type.');
+}
```

Comment:
I'd split in 3 functions, 1 for arrays, 1 for objects, and `computePresentationType`, which I'd rename to something shorter like `getDisplayType`.
It probably needs some simplification, which we can do later